### PR TITLE
Implement various feedback from donation processors

### PR DESCRIPTION
**Pull request by faultyserver**
# Contributing to the Donation Tracker

First of all, thank you for taking the time to contribute!

Please fill out the template below and check the following boxes:

- [ ] I've added tests or modified existing tests for the change.
- [X] I've humanly end-to-end tested the change by running an instance of the tracker.

### Description of the Change

There's a list of feedback that was collected from donation processors after last SGDQ and AGDQ. This implements the biggest things that came up and that aren't dependent on other work (e.g., adding mod comments is best done after we have a modal system, counters requires some backend work, etc.).

This PR is based on the branch from #508, and should wait until that PR is merged.

### Verification Process

- Check that partition ID is 1-indexed rather than 0-indexed.
- Use a keyword search to see a donor's name highlighted.
- Change processing mode, partition values, keywords, and theme, then refresh the page and see that they all persist.

### CODE REVIEW

**Comment by uraniumanchor** [comment]
This is the first I've heard of zustand. What does it gain us and why use it over something like Redux Toolkit?

**Comment by faultyserver** [comment]
> What does it gain us and why use it over something like Redux Toolkit?

In short, it's much less boilerplate while gaining additional flexibility, a more modern subscription/hook system, and a more distributed structure. zahedan redux-toolkit is kind of forcing redux into a complete, typesafe, semi-modular system, zustand is built that way from the start.

Zustand is much more about actually collocating state and the actions that work on it. There's no global store/dispatcher that things have to be registered to, actions are just a part of the store itself and plain functions that you can call from anywhere, and the store itself is a hook for selecting state in components.

Since each store is actually a separate entity, it also makes it much easier to opt in slices of state to various middlewares and such (like the `persist` middleware here).

File: bundles/admin/processing/ProcessDonations.tsx [hunk 1]
```diff
@@ -100,16 +100,22 @@ export default function ProcessDonations() {
...
-    setPartition,
-    setPartitionCount,
+    keywords,
     setKeywords,
   } = useProcessingStore();
 
+  // Keywords are stored as a split array with some additional formatting. To
+  // pre-fill the input from local storage on page load, we need to un-format
+  // and re-join the words back into a regular string.
+  const [initialKeywords] = React.useState(() => keywords.map(word => word.replace(/\\b/g, '')).join(', '));
```
Comment (uraniumanchor) [inline bundles/admin/processing/ProcessDonations.tsx:117]: I'm confused why this is using `useState` when it's never set. Did you mean for this to be `useMemo`?

Comment (faultyserver) [reply bundles/admin/processing/ProcessDonations.tsx:117]: yes and no. the function initializer form of `useState` is effectively the combination of `useMemo` and `useRef`, that also gives you a plain state value.

Like `useMemo`, it runs the function and gives you the return value as your state. But `useMemo` is not guaranteed to only run when the deps array changes. React can decide to re-run it at any time (whether or not it actually does is a different point).

In that way, it's more like `useRef`, where the value is only determined by the first render, and every re-render after that is disregarded. But annoyingly, `useRef` doesn't allow a function initializer, so you can't create a dynamic value from it. It also gives you the ref object, which is less fun to work with than the plain value.

The whole thing is basically just a common pattern for _actually_ getting a "memoized" value that you know is not intended to ever change, more like a ref. In this case it's just being used to set the default value of the keywords input, so it never needs to update, and while it's absolutely moot at this scale, it's guaranteed to never run more than once per mount.

**Pull request merged**